tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -1863,6 +1894,56 @@ bool StitchCPUAnalysis::doBcastOpTileAnalysis(
   return true;
 }
 
+SmallVector<int> getNonSizeOneDims(MemRefType ty) {
+  SmallVector<int> dims;
+  for (int i = 0; i < ty.getRank(); ++i)
+    if (ty.getShape()[i] != 1) dims.push_back(i);
+  return dims;
+}
+
+bool StitchCPUAnalysis::doReshapeOpTileAnalysis(
+    DenseMap<Value, TileInfo>& tilePlan, Operation* op, bool& changed) {
+  Value inValue = op->getOperand(0);
+  Value outValue = cast<lmhlo::LmhloOp>(op).getResultBuffer();
+  // target_shape should not do parallel computation.
+  if (isa<lmhlo::DynamicReshapeOp>(op)) {
+    auto& targetShape = tilePlan[op->getOperand(1)];
+    TileInfo mergedInfo = targetShape;
+    if (!mergedInfo.merge(0)) return false;
+    changed |= targetShape.updateIfNotEqual(mergedInfo);
+  }
+
+  auto propagateTileDims = [&](Value in, TileInfo& inInfo, Value out,
+                               TileInfo& outInfo) {
+    auto inTy = in.getType().cast<MemRefType>();
+    auto inNonSizeOneDims = getNonSizeOneDims(inTy);
+    auto outTy = out.getType().cast<MemRefType>();
+    auto outNonSizeOneDims = getNonSizeOneDims(outTy);
+
+    if (inNonSizeOneDims.size() != outNonSizeOneDims.size()) return false;
+    for (const auto& z : llvm::zip(inNonSizeOneDims, outNonSizeOneDims)) {
+      int inIdx, outIdx;
+      std::tie(inIdx, outIdx) = z;
+      auto outIt = outInfo.tileSizes.find(outIdx);
+      if (outIt != outInfo.tileSizes.end())
+        if (!inInfo.merge(inIdx, outIt->second)) return false;
+    }
+    return true;
+  };
+
+  TileInfo inMergedInfo = tilePlan[inValue];
+  TileInfo outMergedInfo = tilePlan[outValue];
+
+  if (!propagateTileDims(inValue, inMergedInfo, outValue, outMergedInfo) ||
+      !propagateTileDims(outValue, outMergedInfo, inValue, inMergedInfo)) {
+    return false;
+  }
```

Comment:
Return true only when both in->out/out->in are propagatable?

Reply:
Yes. We need the tile plans for the input and output are consistent.